renormalizer/mps/mps.py

```diff
@@ -375,10 +372,10 @@ def expectation(self, mpo, self_conj=None) -> Union[float, complex]:
         else:
             ret_float = False
         environ = Environ(self, mpo, "R", mps_conj=self_conj)
-        l = ones((1, 1, 1))
+        l = xp.ones((1, 1, 1), dtype=self.dtype)
         r = environ.read("R", 1)
         path = self._expectation_path()
-        val = multi_tensor_contract(path, l, self[0], mpo[0], self_conj[0], r).array
+        val = multi_tensor_contract(path, l, self[0], mpo[0], self_conj[0], r)
         if ret_float:
```

Comment:
The `ret_float` is dangerous, only when the mpo is hermitian, the expectation value is real.

Reply:
Currently, we only have hermitian operators. If a non-hermitian operator is used it is a significant change and the user would expect complex output so that the bug can be found without much effort. The reason why it is designed this way is that it avoids annoying warnings issued by `matplotlib` when ploting imaginary numbers. I agree to add an assertion here though.